src/OrleansGetEventStore/GetEventStoreProvider.cs

```diff
@@ -0,0 +1,304 @@
+                    ? ExpectedVersion.NoStream
+                    : expectedVersion.Value - 1
+                : ExpectedVersion.Any;
+
+            if (eventsToSave.Length < pageSize)
+            {
+                try
+                {
+                    await this.Connection.AppendToStreamAsync(streamName, version, eventsToSave);
+                }
+                catch (WrongExpectedVersionException)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                try
+                {
+                    var transaction = await this.Connection.StartTransactionAsync(streamName, version);
+
+                    var position = 0;
+                    while (position < eventsToSave.Length)
+                    {
+                        var pageEvents = eventsToSave.Skip(position).Take(pageSize);
+                        await transaction.WriteAsync(pageEvents);
+                        position += pageSize;
+                    }
+
+                    await transaction.CommitAsync();
+                }
+                catch (WrongExpectedVersionException)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static List<KeyValuePair<Guid, object>> emptyList = new List<KeyValuePair<Guid, object>>();
+
+        public async Task<int> GetVersion(string streamName)
+        {
+            // read the latest event
+            var response = await Connection.ReadEventAsync(streamName, StreamPosition.End, true);
+
+            if (response.Status == EventReadStatus.NotFound
+                || response.Status == EventReadStatus.NoStream
+                || response.Status == EventReadStatus.StreamDeleted)
+            {
+                return 0;
+            }
+
+            return response.Event.Value.Event.EventNumber + 1;
+        }
+
+        public async Task<EventStreamSegment> Load(string streamName, int fromVersion, int? toVersion)
+        {
+            // fix requests with invalid ranges to closest sensible request
+            if (fromVersion < 0)
+                fromVersion = 0;
+            if (toVersion.HasValue && toVersion.Value < fromVersion)
+                toVersion = fromVersion;
+            
+            // special case: requested segment is empty
+            if (toVersion == fromVersion)
+            {
+                // return an empty segment
+                // that is within the current range of the stream, i.e. not larger than latest version
+                var versionThatExists = Math.Min(fromVersion, await GetVersion(streamName));
+                return new EventStreamSegment()
+                {
+                    StreamName = streamName,
+                    FromVersion = versionThatExists,
+                    ToVersion = versionThatExists,
+                    Events = emptyList
+                };
```

Comment:
The spec says that the returned segment must always be a valid segment of the log. There is nothing wrong with asking for an empty sequence.. but we still need to make sure it is a valid subsequence, i.e. not outside the range of the log.